sql/2022/12/inaccurate-sizes-attribute-impact.sql

```diff
@@ -0,0 +1,92 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# See query results here: https://github.com/GoogleChromeLabs/wpp-research/pull/19
+CREATE TEMPORARY FUNCTION getSrcsetSizesAccuracy(payload STRING)
+RETURNS ARRAY<STRUCT<sizesAbsoluteError INT64, sizesRelativeError FLOAT64, wDescriptorAbsoluteError INT64, wDescriptorRelativeError FLOAT64, actualSizesEstimatedWastedLoadedPixels INT64, actualSizesEstimatedWastedLoadedBytes FLOAT64, wastedLoadedPercent FLOAT64>>
+LANGUAGE js AS '''
+try {
+  var $ = JSON.parse(payload);
+  var responsiveImages = JSON.parse($._responsive_images);
+  responsiveImages = responsiveImages['responsive-images'];
+  return responsiveImages.map(({
+    sizesAbsoluteError,
+    sizesRelativeError,
+    wDescriptorAbsoluteError,
+    wDescriptorRelativeError,
+    idealSizesSelectedResourceEstimatedPixels,
+    actualSizesEstimatedWastedLoadedPixels,
+    actualSizesEstimatedWastedLoadedBytes
+  }) => {
+    let wastedLoadedPercent;
+    if ( idealSizesSelectedResourceEstimatedPixels > 0 ) {
+      wastedLoadedPercent = actualSizesEstimatedWastedLoadedPixels / idealSizesSelectedResourceEstimatedPixels;
+    } else {
+      wastedLoadedPercent = null;
+    }
+    return {
+      sizesAbsoluteError,
+      sizesRelativeError,
+      wDescriptorAbsoluteError,
+      wDescriptorRelativeError,
+      actualSizesEstimatedWastedLoadedPixels,
+      actualSizesEstimatedWastedLoadedBytes,
+      wastedLoadedPercent
+    };
+  }
+);
+} catch (e) {
+  return [];
+}
+''';
+
+
+SELECT
+	percentile,
+	client,
+	APPROX_QUANTILES(image.sizesAbsoluteError, 1000)[OFFSET(percentile * 10)] AS sizesAbsoluteError,
+	APPROX_QUANTILES(image.sizesRelativeError, 1000)[OFFSET(percentile * 10)] AS sizesRelativeError,
+	APPROX_QUANTILES(image.wDescriptorAbsoluteError, 1000)[OFFSET(percentile * 10)] AS wDescriptorAbsoluteError,
+	APPROX_QUANTILES(image.wDescriptorRelativeError, 1000)[OFFSET(percentile * 10)] AS wDescriptorRelativeError,
+	APPROX_QUANTILES(image.actualSizesEstimatedWastedLoadedPixels, 1000)[OFFSET(percentile * 10)] AS actualSizesEstimatedWastedLoadedPixels,
+	APPROX_QUANTILES(image.actualSizesEstimatedWastedLoadedBytes, 1000)[OFFSET(percentile * 10)] AS actualSizesEstimatedWastedLoadedBytes,
+	APPROX_QUANTILES(image.wastedLoadedPercent, 1000)[OFFSET(percentile * 10)] AS wastedLoadedPercent
+FROM (
+		SELECT
+			tpages._TABLE_SUFFIX AS client,
+			image
+		FROM
+			`httparchive.pages.2022_10_01_*` AS tpages,
+			UNNEST(getSrcsetSizesAccuracy(payload)) AS image
+				JOIN
+			`httparchive.technologies.2022_10_01_*` AS tech
+			ON
+				tech.url = tpages.url
+		WHERE
+			tpages._TABLE_SUFFIX = tech._TABLE_SUFFIX
+		AND
+			app = 'WordPress'
+		AND
+			category = 'CMS'
```

Comment:
Q. Instead of this query if we use below query what result we get?

```
SELECT
	client,
	image
FROM
	httparchive.all.pages,
	UNNEST(getSrcsetSizesAccuracy(payload)) AS image,
    UNNEST(technologies) as technologies,
    UNNEST(technologies.categories) as category 
WHERE
	technologies.technology = "WordPress"
AND
	category = 'CMS'
AND
    date IN ("2022-10-01")
```

Reply:
@mukeshpanchal27 I'm not sure about the `httparchive.all.pages` table. Can you point out the reference to this version? Thank you!

Reply:
@mukeshpanchal27 I ran both queries to compare them. The format of the results is very different, your query does not result in the data we're looking for. In terms of performance, both queries are extremely expensive to run, but the present query here is still notably cheaper.

The query here consumes 4.2 TB data, while your proposed alternative consumes 7.9 TB of data; so it seems less efficient. Also the query here took about 23 minutes to complete while yours took more than 37 minutes.

The problem in your query is that you use a table that is not month-partitioned. Those tables are _massive_, so we probably shouldn't ever use them.

The original query in this PR comes directly from CMS Almanac, which to me is another indicator we should go with it; so I think we should keep it as is.

cc @vishalkakadiya

Reply:
Thanks @felixarntz for clarification.